File: app/src/main/java/com/shumidub/todoapprealm/realmcontrollers/reportcontroller/ReportRealmController.java
package com.shumidub.todoapprealm.realmcontrollers.reportcontroller;

import com.shumidub.todoapprealm.App;
import com.shumidub.todoapprealm.realmcontrollers.RealmDb;
import com.shumidub.todoapprealm.realmmodel.report.ReportObject;

import java.util.List;

public class ReportRealmController  {


    public static List<ReportObject> getReportList() {
        RealmDb.realm();
        return App.realmFoldersContainer.reportObjectList;
    }


    public static ReportObject getReport(long id) {
        return RealmDb.findById(ReportObject.class, id);
    }


    public static long addReport(String date, int dayCount, String textReport,
                                 int soulRating, int healthRating, int phinanceRating,
                                 int englishRating, int socialRating, int famillyRating,
                                 boolean isWeekReport, int weekNumber) {

        long id = RealmDb.newUniqueId(ReportObject.class);

        RealmDb.write(() -> {
            ReportObject reportObject = RealmDb.realm().createObject(ReportObject.class);
            reportObject.setId(id);
            reportObject.setDate(date);
            reportObject.setCountOfDay(dayCount);
            reportObject.setReportText(textReport);

            reportObject.setSoulRating(soulRating);
            reportObject.setHealthRating(healthRating);
            reportObject.setPhinanceRating(phinanceRating);
            reportObject.setEnglishRating(englishRating);
            reportObject.setSocialRating(socialRating);
            reportObject.setFamillyRating(famillyRating);

            reportObject.setWeekReport(isWeekReport);
            reportObject.setWeekNumber(weekNumber);

            App.realmFoldersContainer.reportObjectList.add(reportObject);
        });
        return id;
    }


    public static void editReport(long id, String date, int dayCount, String textReport,
                                  int soulRating, int healthRating, int phinanceRating,
                                  int englishRating,int socialRating, int famillyRating,
                                  int weekNumber) {
        ReportObject reportObject = getReport(id);
        RealmDb.write(() -> {
            reportObject.setDate(date);
            reportObject.setCountOfDay(dayCount);
            reportObject.setReportText(textReport);

            reportObject.setSoulRating(soulRating);
            reportObject.setHealthRating(healthRating);
            reportObject.setPhinanceRating(phinanceRating);
            reportObject.setEnglishRating(englishRating);
            reportObject.setSocialRating(socialRating);
            reportObject.setFamillyRating(famillyRating);

            reportObject.setWeekNumber(weekNumber);
        });
    }

    public static void delReport(long id) {
        ReportObject reportObject = getReport(id);
        RealmDb.write(() -> {
            App.realmFoldersContainer.reportObjectList.remove(reportObject);
            reportObject.deleteFromRealm();
            RealmDb.realm().where(ReportObject.class).equalTo("id", id).findAll().deleteAllFromRealm();
        });
    }
}

File: app/src/main/java/com/shumidub/todoapprealm/realmcontrollers/reportcontroller/ReportRealmController.kt
package com.shumidub.todoapprealm.realmcontrollers.reportcontroller

import com.shumidub.todoapprealm.App
import com.shumidub.todoapprealm.realmmodel.report.ReportObject
import io.realm.RealmList

object ReportRealmController {

    fun getReportList(): RealmList<ReportObject> {
        App.initRealm()
        return App.realmFoldersContainer?.reportObjectList ?: RealmList()
    }

    fun getReport(id: Long): ReportObject? {
        App.initRealm()
        return App.realm.where(ReportObject::class.java).equalTo("id", id).findFirst()
    }

    fun addReport(
        date: String,
        dayCount: Int,
        textReport: String,
        soulRating: Int,
        healthRating: Int,
        phinanceRating: Int,
        englishRating: Int,
        socialRating: Int,
        famillyRating: Int,
        isWeekReport: Boolean,
        weekNumber: Int,
    ): Long {
        val id = getValidId()
        App.initRealm()
        App.realm.executeTransaction { realm ->
            val report = realm.createObject(ReportObject::class.java).apply {
                this.id = id
                this.date = date
                this.countOfDay = dayCount
                this.reportText = textReport
                this.soulRating = soulRating
                this.healthRating = healthRating
                this.phinanceRating = phinanceRating
                this.englishRating = englishRating
                this.socialRating = socialRating
                this.famillyRating = famillyRating
                this.isWeekReport = isWeekReport
                this.weekNumber = weekNumber
            }
            App.realmFoldersContainer?.reportObjectList?.add(report)
        }
        return id
    }

    fun editReport(
        id: Long,
        date: String,
        dayCount: Int,
        textReport: String,
        soulRating: Int,
        healthRating: Int,
        phinanceRating: Int,
        englishRating: Int,
        socialRating: Int,
        famillyRating: Int,
        weekNumber: Int,
    ) {
        App.initRealm()
        val report = getReport(id) ?: return
        App.realm.executeTransaction {
            report.date = date
            report.countOfDay = dayCount
            report.reportText = textReport
            report.soulRating = soulRating
            report.healthRating = healthRating
            report.phinanceRating = phinanceRating
            report.englishRating = englishRating
            report.socialRating = socialRating
            report.famillyRating = famillyRating
            report.weekNumber = weekNumber
        }
    }

    fun delReport(id: Long) {
        App.initRealm()
        val report = getReport(id) ?: return
        val deletion: () -> Unit = {
            App.realmFoldersContainer?.reportObjectList?.remove(report)
            report.deleteFromRealm()
            App.realm.where(ReportObject::class.java)
                .equalTo("id", id)
                .findAll()
                .deleteAllFromRealm()
        }
        if (App.realm.isInTransaction) deletion() else App.realm.executeTransaction { deletion() }
    }

    private fun getValidId(): Long {
        App.initRealm()
        var id = System.currentTimeMillis()
        while (App.realm.where(ReportObject::class.java).equalTo("id", id).findFirst() != null) {
            id++
        }
        return id
    }
}
